### PR TITLE
Typmod handling for functions expression with numeric.

### PR DESCRIPTION
**Pull request by Yvinayak07**
### Issue
In `resolve_numeric_typmod_from_exp` we missed the handling for typmod calculation for some mathematical functions  in function expression like `ROUND(), POWER(), CEILING(), FLOOR(), SIGN(), ABS(), DEGREES(), RADAINS(), SCOPE_IDENTITY(), IDENT_SEED(), IDENT_INCR(), IDENT_CURRENT()`.

### 2. Changes made to fix the issues

Added handling for `ROUND(), POWER(), CEILING(), FLOOR(), SIGN(), ABS(), DEGREES(), RADAINS(), SCOPE_IDENTITY(), IDENT_SEED(), IDENT_INCR(), IDENT_CURRENT()`  mathematical functions in `T_FuncExpr` node

### Issues Resolved

Task: BABEL-5666

Authored-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)
Signed-off-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)

### Test Scenarios Covered ###
* **Use case based -** YES

* **Boundary conditions -** YES

* **Arbitrary inputs -** YES

* **Negative test cases -** YES

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 1]
```diff
@@ -1679,24 +1711,81 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
...
 				 * If the following conditions are met then we will recursively find typmod from arg.
 				 * 1) rettypmod == -1 means unable to find typmod till now.
 				 * 2) check if only one args and then is that castable to numeric.
 				 */
 				if (rettypmod == -1 &&
-					list_length(func->args) == 1 &&
-					is_numeric_cast(func_oid))
+					list_length(func->args) >= 1 &&
+					(is_numeric_cast(func_oid) || 
+					(funcName && is_mathematical_function(funcName))))
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Why?
Can't we just move the entire logic down?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 2]
```diff
@@ -1679,24 +1711,81 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
...
+					{
+						if ((*common_utility_plugin_ptr->is_tsql_smallmoney_datatype) (func->funcresulttype))
+							fixlen_default_typmod = TSQL_MONEY_TYPMOD;
+
+						pfree(funcName);
+						return fixlen_default_typmod;
+					}
+					else
+					{
+						if ((strlen(funcName) == 7 && (strncmp(funcName, "ceiling", 7) == 0)) ||
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Sacrificing efficiency and readability by string comparison twice over calling fixed Len datatype or generic numeric for  each.
I would choose the latter only for efficiency.

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 1 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Moved this change down so that we call `is_mathematical_function` once.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 3]
```diff
@@ -1697,6 +1727,74 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
 					return rettypmod;
 				}
 
+				/* We will find typmod for mathematical functions. */
+				funcName = get_func_name(func_oid);
+				if (funcName && is_mathematical_function(funcName))
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: funcName == NULL check is already there in is_mathematical_function function, no need of adding it here again.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 4]
```diff
@@ -1697,6 +1727,74 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
...
+					{
+						pfree(funcName);
+						return fixsize_default_typmod;
+					}
+					else
+					{
+						uint8_t		precision = 0,
+								scale = 0;
+
+						if (list_length(func->args) >= 1)
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: why do we need this check ? and If we need this check how are we handling the other case where list_length(func->args) < 1 ?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 5]
```diff
@@ -1697,6 +1727,74 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
...
+							scale = (rettypmod - VARHDRSZ) & 0xffff;
+							precision = ((rettypmod - VARHDRSZ) >> 16) & 0xffff;
+						}
+						
+						if ((strlen(funcName) == 7 && (strncmp(funcName, "ceiling", 7) == 0)) ||
+								(strlen(funcName) == 5 && (strncmp(funcName, "floor", 5) == 0)))
+						{
+							/* for ceiling and floor functions, we return 0 scale */
+							scale = 0;
+						}
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: TSQL Documentation says that the return typmod should be `38, s`

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 6]
```diff
@@ -1697,6 +1727,74 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
...
+							/*
+							 * For abs and power functions, precision as 38.
+							 */
+							precision = tds_default_numeric_precision;
+						}
+						rettypmod = ((precision << 16) | scale) + VARHDRSZ;
+					}
+				}
+
+				pfree(funcName);
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: we should have a sanity check in case if funcName is NULL

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 2 above)
Comment (rohit01010) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: +1

**Review by rohit01010 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 4 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: The recursive call here is for calculating typmod from first arguments from the functions like `power, round` etc. so we need to check if arguments are greater then 0. Their are some functions that have 0 arguments like `scope_identity, ident_seed, ident_incr, ident_current` these have fixed typmod i.e (38,0).

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 3 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Addressed.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 5 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: But their is behaviour difference in execution. confirmed using information_schema.columns:
```
1> select ceiling(cast(1 as numeric(12,3))) col1,
    radians(cast(1 as numeric(14,2))) col2, 
   degrees(cast(1 as numeric(14,2))) col3,
   floor(cast(1 as numeric(14,2))) col4 into t1
2> go

(1 rows affected)
1> select COLUMN_NAME, numeric_scale, numeric_precision
   from information_schema.columns where table_name = 't1'
2> go
COLUMN_NAME          numeric_scale numeric_precision
-------------------- ------------- -----------------
col1                 0                12  -- ceiling
col2                18                38  -- radians
col3                18                38  -- degrees
col4                 0                14  -- floor

(4 rows affected)
```

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 6 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: addressed.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 2 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: I have changed the code where we only do string string matching once.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 7]
```diff
@@ -1697,10 +1721,132 @@ resolve_numeric_typmod_from_exp(Plan *plan, Node *expr, bool *found)
...
+								(strlen(funcName) == 10 && (strncmp(funcName, "ident_seed", 10) == 0)) ||
+								(strlen(funcName) == 10 && (strncmp(funcName, "ident_incr", 10) == 0)) ||
+								(strlen(funcName) == 13 && (strncmp(funcName, "ident_current", 13) == 0)))
+					{
+						int32 fixsize_default_typmod = get_default_typmod_for_fixedsize_dataypes(func->funcresulttype);
+						if (fixsize_default_typmod != -1)
+						{
+							pfree(funcName);
+							return fixsize_default_typmod;
+						}
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: This might not be required for functions scope_identity, ident_seed, ident_incr, ident_incr, ident_current

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 8]
```diff
@@ -1234,6 +1234,28 @@ get_default_typmod_for_fixedsize_dataypes(Oid resulttype)
...
+
+	if (list_length(args) >= 1)
+	{
+		arg = linitial(args);
+		rettypmod = resolve_numeric_typmod_from_exp(plan, arg, found_typmod);
+	}
+	if (rettypmod != -1)
+		return rettypmod;
+
+	return -1;
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: NIT: we can directly return rettypmod, no need of if check.

Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: addressed.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 7 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Yes as these functions have fixed return typmod. addressed.

**Review by rohit01010 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 9]
```diff
@@ -1218,9 +1218,9 @@ is_numeric_datatype(Oid typid)
 static int32
 get_default_typmod_for_fixedsize_dataypes(Oid resulttype)
 {
-	if (resulttype == INT4OID)
+	if (resulttype == INT4OID || (*common_utility_plugin_ptr->is_tsql_int_datatype)(resulttype))
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:1224]: TSQL Int type? Please explain

File: contrib/babelfishpg_common/src/typecode.c [hunk 10]
```diff
@@ -521,6 +523,22 @@ is_tsql_tinyint_datatype(Oid oid)
...
+	return tsql_int_oid == oid;
+}
+
+bool
+is_tsql_bigint_datatype(Oid oid)
+{
+	if (tsql_bigint_oid == InvalidOid)
+		tsql_bigint_oid = lookup_tsql_datatype_oid("bigint");
+	return tsql_bigint_oid == oid;
+}
```
Comment (KushaalShroff) [inline contrib/babelfishpg_common/src/typecode.c:543]: Typecodes mention that these non sys types for babelfish:
```
	{0, 0, "int8", "bigint", 4, 13, 1},
	{0, 0, "int4", "int", 4, 14, 1},
	{0, 0, "int2", "smallint", 4, 15, 1},
```
It seems counter intuitive to introduce them

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 9 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:1224]: We have some round functions that have TSQL int and TSQL bigint as return type:
```
1> select proname, pronargs,proargtypes, prorettype from pg_proc where proname like 'round'
2> go
proname                                                                                                                                                                                                                                                          pronargs proargtypes                                                                                                                                                                                                                                                      prorettype 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- -------- ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- -----------
round                                                                                                                                                                                                                                                                   1 701                                                                                                                                                                                                                                                                      701
round                                                                                                                                                                                                                                                                   2 1700 23                                                                                                                                                                                                                                                                 1700
round                                                                                                                                                                                                                                                                   1 1700                                                                                                                                                                                                                                                                    1700
round                                                                                                                                                                                                                                                                   2 1700 23                                                                                                                                                                                                                                                                17132
round                                                                                                                                                                                                                                                                   3 1700 23 23                                                                                                                                                                                                                                                             17132
round                                                                                                                                                                                                                                                                   2 23 23                                                                                                                                                                                                                                                                  17124
round                                                                                                                                                                                                                                                                   3 23 23 23                                                                                                                                                                                                                                                               17124
round                                                                                                                                                                                                                                                                   2 17126 23                                                                                                                                                                                                                                                               17126
round                                                                                                                                                                                                                                                                   3 17126 23 23                                                                                                                                                                                                                                                            17126
round                                                                                                                                                                                                                                                                   2 16399 23                                                                                                                                                                                                                                                               16413
round                                                                                                                                                                                                                                                                   3 16399 23 23                                                                                                                                                                                                                                                            16413
round                                                                                                                                                                                                                                                                   2 17130 23                                                                                                                                                                                                                                                               17130
round                                                                                                                                                                                                                                                                   3 17130 23 23                                                                                                                                                                                                                                                            17130

(13 rows affected)
```

File: contrib/babelfishpg_common/src/typecode.c (hunk 10 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_common/src/typecode.c:543]: It was because of the reason mentioned [above](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3868#discussion_r2179152656).

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 9 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:1224]: `17124` is for TSQL int and `17126` is for TSQL bigint.

**Review by KushaalShroff — DISMISSED** [review]

**Review by rohit01010 — APPROVED** [review]

**Review by jsudrik — APPROVED** [review]
Approved.
